packages/admin-panel-server/src/viz-builder/dashboardVisualisation/reportConfigValidator.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup, yupUtils } from '@tupaia/utils';
+
+// TODO avoid duplication with report-server
+
+const periodTypeValidator = yup.mixed().oneOf(['day', 'week', 'month', 'quarter', 'year']);
+
+const createDataSourceValidator = (sourceType: 'dataElement' | 'dataGroup') => {
+  const otherSourceKey = sourceType === 'dataElement' ? 'dataGroups' : 'dataElements';
+
+  return yup
+    .array()
+    .of(yup.string())
+    .when(['$testData', otherSourceKey], {
+      is: ($testData: unknown, otherDataSource: string[]) =>
+        !$testData && (!otherDataSource || otherDataSource.length === 0),
+      then: yup.array().of(yup.string()).required('Requires "dataGroups" or "dataElements"').min(1),
+      otherwise: yup.array().of(yup.string()),
+    });
+};
+
+const dataElementValidator = createDataSourceValidator('dataElement');
+const dataGroupValidator = createDataSourceValidator('dataGroup');
+
+const dateSpecsValidator = yupUtils.polymorphic({
+  object: yup
+    .object()
+    .shape({
+      unit: periodTypeValidator.required(),
+      offset: yup.number(),
+      modifier: yup.mixed().oneOf(['start_of', 'end_of']),
+      modifierUnit: periodTypeValidator,
+    })
+    .default(undefined),
+  string: yup.string().min(4),
+});
+
+export const reportConfigValidator = yup.object().shape({
+  fetch: yup.object().shape(
+    {
+      dataElements: dataElementValidator,
+      dataGroups: dataGroupValidator,
+      aggregations: yup.array().required(),
+      startDate: dateSpecsValidator,
+      endDate: dateSpecsValidator,
+    },
+    [['dataElements', 'dataGroups']],
+  ),
+  transform: yup.array().required(),
+  output: yup.object(),
```

Comment:
We can extract and use existing validators from `transform.ts` and `output.ts` here. Similarly for `fetch.aggregations` (out of scope for this PR)

Reply:
I think thorough validation is valuable here, because it will prevent users from importing reports that will end up throwing errors